program/src/processor.rs

```diff
@@ -45,53 +45,63 @@ fn calculate_proposal_vote_threshold(stake: u64, total_stake: u64) -> Result<u64
     // Ensure the stake account is owned by the Paladin Stake program.
     if stake_info.owner != &paladin_stake_program::id() {
         return Err(ProgramError::InvalidAccountOwner);
     }
 
-    let data = stake_info.try_borrow_data()?;
-    let state =
-        bytemuck::try_from_bytes::<Stake>(&data).map_err(|_| ProgramError::InvalidAccountData)?;
+    let stake = {
+        let data = stake_info.try_borrow_data()?;
+        let state = bytemuck::try_from_bytes::<Stake>(&data)
+            .map_err(|_| ProgramError::InvalidAccountData)?;
 
-    // Ensure the stake account is initialized.
-    if !state.is_initialized() {
-        return Err(ProgramError::UninitializedAccount);
-    }
+        // Ensure the stake account is initialized.
+        if !state.is_initialized() {
+            return Err(ProgramError::UninitializedAccount);
+        }
 
-    // Ensure the stake account belongs to the authority.
-    if state.authority != *authority_key {
-        return Err(ProgramError::IncorrectAuthority);
-    }
+        // Ensure the stake account belongs to the authority.
+        if state.authority != *authority_key {
+            return Err(ProgramError::IncorrectAuthority);
+        }
 
-    // Return the currently active stake amount.
-    Ok(state.amount)
-}
+        // Ensure the stake account has the correct address derived from the
+        // validator vote account and the stake config account.
+        if stake_info.key
+            != &find_stake_pda(
+                &state.validator_vote,
+                stake_config_info.key,
+                &paladin_stake_program::id(),
+            )
+            .0
+        {
+            return Err(PaladinGovernanceError::StakeConfigMismatch.into());
+        }
 
-fn get_total_stake_checked(
-    _authority_key: &Pubkey,
-    _stake_key: &Pubkey,
-    stake_config_info: &AccountInfo,
-) -> Result<u64, ProgramError> {
-    // Ensure the config account is owned by the Paladin Stake program.
+        state.amount
+    };
+
+    // Ensure the stake config account is owned by the Paladin Stake program.
     if stake_config_info.owner != &paladin_stake_program::id() {
         return Err(ProgramError::InvalidAccountOwner);
     }
 
-    let data = stake_config_info.try_borrow_data()?;
-    let state = bytemuck::try_from_bytes::<StakeConfig>(&data)
-        .map_err(|_| ProgramError::InvalidAccountData)?;
+    let total_stake = {
+        let data = stake_config_info.try_borrow_data()?;
+        let state = bytemuck::try_from_bytes::<StakeConfig>(&data)
+            .map_err(|_| ProgramError::InvalidAccountData)?;
 
-    // Ensure the config account is initialized.
-    if !state.is_initialized() {
-        return Err(ProgramError::UninitializedAccount);
-    }
+        // Ensure the config account is initialized.
+        if !state.is_initialized() {
+            return Err(ProgramError::UninitializedAccount);
+        }
 
-    // TODO: How to check the config account is the correct one?
```

Comment:
I'm not sure if this is addressed -- we might need the governance too, to make sure the correct stake config is provided

Reply:
I thought we said the PDA check was sufficient? We validate that the authority who signs is in fact the authority on the stake account, and that stake account's stored vote address + config derives to the stake account's address.

Reply:
definitely the stake account is good, but are we sure it's the correct stake config?

Reply:
We know it's the correct stake config for the stake account, but we don't know it's the correct config for the governance config. Let me add a TODO for when the governance config bit is addressed.

Reply:
Ok, so do you want to add the TODOs and then merge this?

Reply:
Yep. Added.